index.bs

```diff
@@ -1874,6 +1882,585 @@ A few abstract operations are used in this specification for utility purposes. W
+  1. If _x_ does not have a [[controlledReadableByteStream]] internal slot, return *false*.
+  1. Return *true*.
+</pre>
+
+<h4 id="is-readable-byte-stream-locked" aoid="IsReadableByteStreamLocked">IsReadableByteStreamLocked ( stream )</h4>
+
+<pre is="emu-alg">
+  1. Assert: IsReadableByteStream(_stream_) is *true*.
+  1. If _stream_@[[reader]] is *undefined*, return *false*.
+  1. Return *true*.
+</pre>
+
+<h4 id="is-readable-byte-stream-reader" aoid="IsReadableByteStreamReader">IsReadableByteStreamReader ( x )</h4>
+
+<pre is="emu-alg">
+  1. If Type(_x_) is not Object, return *false*.
+  1. If _x_ does not have an [[ownerReadableByteStream]] internal slot, return *false*.
+  1. Return *true*.
+</pre>
+
+<h4 id="read-from-readable-byte-stream-reader" aoid="ReadFromReadableByteStreamReader">ReadFromReadableByteStreamReader ( reader )</h4>
+
+<pre is="emu-alg">
+  1. If _reader_@[[state]] is "closed", return a new promise resolved with CreateIterResultObject(*undefined*, *true*).
+  1. If _reader_@[[state]] is "errored", return a new promise rejected with _reader_@[[storedError]].
+  1. Assert: _reader_@[[ownerReadableByteStream]] is not *undefined*.
+  1. Assert: _reader_@[[ownerReadableByteStream]]@[[state]] is "readable".
+  1. Let _readRequestPromise_ be a new promise.
+  1. Append _readRequestPromise_ as the last element of _reader_@[[readRequests]].
+  1. Let _owner_ be _reader_@[[ownerReadableByteStream]].
+  1. Let _pullResult_ be Invoke(_owner_, "pull").
+  1. If _pullResult_ is an abrupt completion,
+    1. If _owner_@[[state]] is "readable", ErrorReadableByteStream(_owner_, _pullResult_.[[value]]).
+  1. Return _readRequestPromise_.
+</pre>
+
+<h4 id="release-readable-byte-stream-reader" aoid="ReleaseReadableByteStreamReader">ReleaseReadableByteStreamReader ( reader )</h4>
+
+<pre is="emu-alg">
+  1. Assert: _reader_@[[ownerReadableByteStream]] is not *undefined*.
+  1. If _reader_@[[ownerReadableByteStream]]@[[state]] is "errored",
+    1. Set _reader_@[[state]] to "errored".
+    1. Let _e_ be _reader_@[[ownerReadableByteStream]]@[[storedError]].
+    1. Set _reader_@[[storedError]] to _e_.
+    1. Reject _reader_@[[closedPromise]] with _e_.
+    1. Repeat for each _readRequestPromise_ that is an element of _reader_@[[readRequests]],
+      1. Reject _readRequestPromise_ with _e_.
+  1. Otherwise,
+    1. Set _reader_@[[state]] to "closed".
+    1. Resolve _reader_@[[closedPromise]] with *undefined*.
+    1. Repeat for each _readRequestPromise_ that is an element of _reader_@[[readRequests]],
+      1. Resolve _readRequestPromise_ with CreateIterResultObject(*undefined*, *true*).
+  1. Set _reader_@[[readRequests]] to a new empty List.
+  1. Set _reader_@[[ownerReadableByteStream]]@[[reader]] to *undefined*.
+  1. Set _reader_@[[ownerReadableByteStream]] to *undefined*.
+</pre>
+
+<h4 id="tee-readable-byte-stream" aoid="TeeReadableByteStream">TeeReadableByteStream ( stream, shouldClone )</h4>
```

Comment:
Hmm, this is complicated. With the queues and such. Could you summarize for me how it works? It looks similar to TeeReadableStream but I am betting there are important differences.